src/McpServerService.cs

```diff
@@ -0,0 +1,168 @@
+        private bool _isPackageDataReady = false;
+        private McpServer _server;
+
+        /// <summary>
+        /// Initializes a new instance of the McpServerService.
+        /// </summary>
+        public McpServerService()
+        {
+            _cancellationTokenSource = new CancellationTokenSource();
+            _packageLoadedEvent = new ManualResetEventSlim(false);
+        }
+
+        /// <summary>
+        /// Starts the MCP server and waits for package data to be loaded.
+        /// This method should be called after the application has started loading packages.
+        /// </summary>
+        /// <param name="cancellationToken">Token to cancel the server operation</param>
+        /// <returns>A task representing the server operation</returns>
+        public async Task StartAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                // Wait for the main window to finish loading packages
+                await WaitForPackageDataAsync(cancellationToken);
+
+                // Create and start the MCP server
+                _server = new McpServer("view-appxpackage-mcp", "1.0.0");
+                
+                // Register this service instance to handle tool calls
+                _server.AddToolProvider(this);
+
+                // Start the server - this will block until cancellation is requested
+                await _server.RunAsync(cancellationToken);
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected when cancellation is requested
+            }
+            catch (Exception ex)
+            {
+                // Log any unexpected errors
+                DebugLog.Append($"MCP Server error: {ex.Message}");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Stops the MCP server gracefully.
+        /// </summary>
+        public void Stop()
+        {
+            _cancellationTokenSource.Cancel();
+            _server?.Dispose();
+            _packageLoadedEvent.Dispose();
+        }
+
+        /// <summary>
+        /// Waits for the package data to be loaded and ready for serving.
+        /// This ensures we don't serve stale or incomplete data.
+        /// </summary>
+        /// <param name="cancellationToken">Token to cancel the wait operation</param>
+        /// <returns>A task that completes when package data is ready</returns>
+        private async Task WaitForPackageDataAsync(CancellationToken cancellationToken)
+        {
+            // Check if MainWindow instance exists and IsLoading is false
+            while (MainWindow.Instance?.IsLoading != false)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException();
+                }
+
+                // Wait a short period before checking again
+                await Task.Delay(100, cancellationToken);
```

Comment:
Rather than a sleep, listen to the MainWindow.PropertyChanged event for the "IsLoading" property to change to false. It's not necessary to do a null check on MainWindow.Instance; it's guaranteed to be non-null when this method is called